lib/nxt_pipeline/pipeline.rb

```diff
@@ -115,22 +119,75 @@ def on_errors(*errors, halt_on_error: true, &callback)
 
     alias :on_error :on_errors
 
-    def before_execute(&callback)
-      self.before_execute_callback = callback
+    def before_step(&block)
+      callbacks.resolve!(:step, :before) << block
+    end
+
+    def after_step(&block)
+      callbacks.resolve!(:step, :after) << block
+    end
+
+    def around_step(&block)
+      callbacks.resolve!(:step, :around) << block
     end
 
-    def after_execute(&callback)
-      self.after_execute_callback = callback
+    def before_execution(&block)
+      callbacks.resolve!(:execution, :before) << block
+    end
+
+    def after_execution(&block)
+      callbacks.resolve!(:execution, :after) << block
+    end
+
+    def around_execution(&block)
+      callbacks.resolve!(:execution, :around) << block
     end
 
     private
 
+    def run_callbacks(type, kind, changeset)
+      callbacks.resolve!(type, kind).each do |callback|
+        run_callback(callback, changeset)
+      end
+    end
+
+    def run_around_callbacks(type, args, &execution)
+      around_callbacks = callbacks.resolve!(type, :around)
+      return execution.call unless around_callbacks.any?
+
+      callback_chain = around_callbacks.reverse.inject(execution) do |previous, callback|
```

Comment:
Nope that's not quite it. This thing is indeed crazy code and is nesting callbacks into each other. If you think about how around callbacks work, then you will notice that the second around callback runs inside of the first around callback and so on. So that thing here returns a proc that runs all callbacks inside each other and then executes the step / or execution. It's not that easy to understand, but it lies in the nature of the problem. I challenge you to implement around callbacks without looking at this :-D